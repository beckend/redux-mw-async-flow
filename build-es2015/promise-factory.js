"use strict";
const Bluebird = require("bluebird");
exports.createPromise = () => {
    let resolve;
    let reject;
    const promise = new Bluebird((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return {
        promise,
        resolve: resolve,
        reject: reject
    };
};
//# sourceMappingURL=promise-factory.js.map