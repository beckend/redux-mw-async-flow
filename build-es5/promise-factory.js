"use strict";
var Bluebird = require("bluebird");
exports.createPromise = function () {
    var resolve;
    var reject;
    var promise = new Bluebird(function (res, rej) {
        resolve = res;
        reject = rej;
    });
    return {
        promise: promise,
        resolve: resolve,
        reject: reject,
    };
};
//# sourceMappingURL=promise-factory.js.map