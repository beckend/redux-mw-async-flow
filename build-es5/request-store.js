"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lGet = require("lodash.get");
exports.REQUEST_KEY_PROMISE = 'promise';
exports.REQUEST_KEY_RESOLVEFN = 'resolve';
exports.REQUEST_KEY_REJECTFN = 'reject';
var RequestStore = /** @class */ (function () {
    function RequestStore() {
        this.store = {};
    }
    RequestStore.prototype.add = function (keyName, payload) {
        this.store[keyName] = payload;
    };
    RequestStore.prototype.delete = function (keyName) {
        delete this.store[keyName];
    };
    RequestStore.prototype.get = function (keyName) {
        return this.store[keyName];
    };
    RequestStore.prototype.resolve = function (keyName, payload) {
        var promiseFn = lGet(this.store, [keyName, exports.REQUEST_KEY_RESOLVEFN]);
        if (promiseFn) {
            promiseFn(payload);
        }
        else {
            // tslint:disable-next-line: no-console
            console.warn(keyName + " - was not found in request store and was not resolved.");
        }
    };
    RequestStore.prototype.reject = function (keyName, payload) {
        var promiseFn = lGet(this.store, [keyName, exports.REQUEST_KEY_REJECTFN]);
        if (promiseFn) {
            promiseFn(payload);
        }
        else {
            // tslint:disable-next-line: no-console
            console.warn(keyName + " - was not found in request store and was not rejected.");
        }
    };
    return RequestStore;
}());
exports.RequestStore = RequestStore;
//# sourceMappingURL=request-store.js.map