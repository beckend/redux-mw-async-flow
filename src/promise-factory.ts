import * as Bluebird from 'bluebird';

export type TResolveFn<TPayload> = (thenableOrResult?: TPayload | Bluebird.Thenable<TPayload>) => void;
export type TRejectFn = (error?: Error | any) => void;

export const createPromise = <TPayload>() => {
  let resolve: any;
  let reject: any;
  const promise = new Bluebird<TPayload>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  return {
    promise,
    resolve: resolve as TResolveFn<TPayload>,
    reject: reject as TRejectFn,
  };
};
