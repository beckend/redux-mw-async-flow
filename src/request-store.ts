/* tslint:disable: no-reserved-keywords */
/**
 * Store promise, resolve, reject outside of action
 */
import * as Bluebird from 'bluebird';
import { TRejectFn, TResolveFn } from './promise-factory';
import lGet = require('lodash.get');

export const REQUEST_KEY_PROMISE = 'promise';
export const REQUEST_KEY_RESOLVEFN = 'resolve';
export const REQUEST_KEY_REJECTFN = 'reject';
export interface IRequestMap<TPayload> {
  readonly promise: Bluebird<TPayload>;
  readonly resolve: TResolveFn<TPayload>;
  readonly reject: TRejectFn;
}
export interface IRequestStore {
  [requestName: string]: IRequestMap<any>;
}
export class RequestStore {

  public store: IRequestStore = {};

  public add(keyName: string, payload: IRequestMap<any>) {
    this.store[keyName] = payload;
  }

  public delete(keyName: string) {
    delete this.store[keyName];
  }

  public get(keyName: string) {
    return this.store[keyName];
  }

  public resolve<TPayload>(keyName: string, payload: TPayload) {
    const promiseFn = lGet<TResolveFn<TPayload>>(this.store, [keyName, REQUEST_KEY_RESOLVEFN]);
    if (promiseFn) {
      promiseFn(payload);
    } else {
      console.warn(`${keyName} - was not found in request store and was not resolved.`);
    }
  }

  public reject<TPayload>(keyName: string, payload: TPayload) {
    const promiseFn = lGet<TResolveFn<TPayload>>(this.store, [keyName, REQUEST_KEY_REJECTFN]);
    if (promiseFn) {
      promiseFn(payload);
    } else {
      console.warn(`${keyName} - was not found in request store and was not rejected.`);
    }
  }

}
