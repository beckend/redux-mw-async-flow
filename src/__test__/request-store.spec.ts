/* tslint:disable: no-backbone-get-set-outside-model */
import { RequestStore } from '../request-store';
import { createPromise } from '../promise-factory';

describe('RequestStore', () => {
  let requestStore: RequestStore;

  beforeEach(() => {
    requestStore = new RequestStore();
  });

  it('has not request initially', () => {
    expect(requestStore.store)
      .toMatchSnapshot();
  });

  it('add', () => {
    requestStore.add('requestKey1', { one: 1 } as any);
    requestStore.add('requestKey2', { one: 1 } as any);
    requestStore.add('requestKey32321', { one: 1 } as any);
    expect(requestStore.store)
      .toMatchSnapshot();
  });

  it('get', () => {
    const requestKey = 'myKey';
    const payload: any = { one: 1 };
    requestStore.add(requestKey, payload);
    expect(requestStore.get(requestKey))
      .toEqual(payload);
    expect(requestStore.get('does-not-exist'))
      .toBeFalsy();
  });

  it('delete', () => {
    const requestKey = 'myKey';
    requestStore.add(requestKey, { one: 1 } as any);
    expect(requestStore.store)
      .toMatchSnapshot();
    requestStore.delete(requestKey);
    requestStore.delete('does not exist');
    expect(requestStore.store)
      .toMatchSnapshot();
  });

  it('resolve', () => {
    const requestKey = 'myKey';
    const resolvePayload = { data: [1, 2, 3] };
    requestStore.add(requestKey, createPromise<any>());
    requestStore.resolve(requestKey, resolvePayload);
    requestStore.resolve('does not exist', resolvePayload);
    expect(requestStore.store)
      .toMatchSnapshot();
    expect(requestStore.get(requestKey).promise.value())
      .toEqual(resolvePayload);
  });

  it('reject', () => {
    const requestKey = 'myKey';
    const rejectPayload = { data: [1, 2, 3] };
    requestStore.add(requestKey, createPromise<any>());
    requestStore.reject(requestKey, rejectPayload);
    requestStore.reject('does not exist', rejectPayload);
    expect(requestStore.store)
      .toMatchSnapshot();
    expect(requestStore.get(requestKey).promise.reason())
      .toEqual(rejectPayload);
  });
});
