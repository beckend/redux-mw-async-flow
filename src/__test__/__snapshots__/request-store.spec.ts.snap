// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestStore add 1`] = `
Object {
  "requestKey1": Object {
    "one": 1,
  },
  "requestKey2": Object {
    "one": 1,
  },
  "requestKey32321": Object {
    "one": 1,
  },
}
`;

exports[`RequestStore delete 1`] = `
Object {
  "myKey": Object {
    "one": 1,
  },
}
`;

exports[`RequestStore delete 2`] = `Object {}`;

exports[`RequestStore has not requests initially 1`] = `Object {}`;

exports[`RequestStore reject 1`] = `
Object {
  "myKey": Object {
    "promise": Object {
      "fulfillmentValue": undefined,
      "isFulfilled": false,
      "isRejected": true,
      "rejectionReason": Object {
        "data": Array [
          1,
          2,
          3,
        ],
      },
    },
    "reject": [Function],
    "resolve": [Function],
  },
}
`;

exports[`RequestStore resolve 1`] = `
Object {
  "myKey": Object {
    "promise": Object {
      "fulfillmentValue": Object {
        "data": Array [
          1,
          2,
          3,
        ],
      },
      "isFulfilled": true,
      "isRejected": false,
      "rejectionReason": undefined,
    },
    "reject": [Function],
    "resolve": [Function],
  },
}
`;
